src/main/java/org/mvel2/integration/impl/IndexVariableResolver.java

```diff
@@ -31,6 +37,27 @@ public Object getValue() {
   }
 
   public void setValue(Object value) {
-    vars[indexPos] = value;
+    initializeType(value); // Initialize type if not yet done
+
+    if (type == Object.class || type == null) {
+      // IndexVariableResolver doesn't know the original declared type, so cannot deal with polymorphism
+      // Just accept the value
+      vars[indexPos] = value;
+    } else {
+      // Coerce numeric types
+      vars[indexPos] = handleTypeCoercion(type, value);
+    }
```

Comment:
This PR makes IndexVariableResolver able to coerce numeric values.

Note that IndexVariableResolver doesn't know the original declared type, so cannot deal with polymorphism, so it just accept the value (= original behavior).

If we want to do type check, we would need to change `SimpleVariableSpaceModel` to store type information (or maybe creating a new class e.g. `TypedVariableSpaceModel`) . So we would also need to change drools side. I wonder if we want to do such a large change for this issue.

Reply:
I totally agree on the fact that it doesn't make any sense to develop any change bigger than this for such a corner case issue. Good job!